dbms/src/Storages/KVStore/MultiRaft/Disagg/CheckpointIngestInfo.cpp

```diff
@@ -0,0 +1,258 @@
+    {
+        // The region should be persisted in local, although it's the first peer of this region in this store.
+        // Otherwise it could be uploaded and then overwritten.
+        RegionPersister::RegionCacheWriteElement region_buffer;
+        RegionPersister::computeRegionWriteBuffer(*region, region_buffer);
+        auto & [region_id, buffer, region_size, applied_index] = region_buffer;
+        if (region->isPendingRemove())
+        {
+            // A pending remove region should not be selected as candidate.
+            LOG_ERROR(DB::Logger::get(), "candidate region {} is pending remove", region->toString(false));
+            return;
+        }
+        auto read_buf = buffer.tryGetReadBuffer();
+        RUNTIME_CHECK_MSG(read_buf != nullptr, "failed to gen buffer for region {}", region->toString(true));
+        RUNTIME_CHECK_MSG(buffer.count() == region_size, "buffer {} != region_size {}", buffer.count(), region_size);
+        auto page_id
+            = UniversalPageIdFormat::toLocalKVPrefix(UniversalPageIdFormat::LocalKVKeyType::FAPIngestRegion, region_id);
+        wb.putPage(UniversalPageId(page_id.data(), page_id.size()), 0, read_buf, region_size);
+    }
+    {
+        size_t data_size = 0;
+        uint64_t count = restored_segments.size();
+        MemoryWriteBuffer buffer;
+        data_size += writeBinary2(count, buffer);
+        for (auto & restored_segment : restored_segments)
+        {
+            data_size += writeBinary2(restored_segment->segmentId(), buffer);
+        }
+        data_size += writeBinary2<UInt64>(remote_store_id, buffer);
+        auto page_id = UniversalPageIdFormat::toLocalKVPrefix(
+            UniversalPageIdFormat::LocalKVKeyType::FAPIngestSegments,
+            region_id);
+        auto read_buf = buffer.tryGetReadBuffer();
+        RUNTIME_CHECK_MSG(buffer.count() == data_size, "buffer {} != data_size {}", buffer.count(), data_size);
+        wb.putPage(UniversalPageId(page_id.data(), page_id.size()), 0, read_buf, data_size);
+    }
+    uni_ps->write(std::move(wb), DB::PS::V3::PageType::Local, nullptr);
+    LOG_INFO(
+        DB::Logger::get(),
+        "Successfully persist CheckpointIngestInfo, region_id={} peer_id={} remote_store_id={} region={}",
+        region_id,
+        peer_id,
+        remote_store_id,
+        region->getDebugString());
+}
+
+void CheckpointIngestInfo::removeFromPS()
+{
+    auto uni_ps = tmt.getContext().getWriteNodePageStorage();
+    UniversalWriteBatch del_batch;
+    del_batch.delPage(
+        UniversalPageIdFormat::toLocalKVPrefix(UniversalPageIdFormat::LocalKVKeyType::FAPIngestRegion, region_id));
+    del_batch.delPage(
+        UniversalPageIdFormat::toLocalKVPrefix(UniversalPageIdFormat::LocalKVKeyType::FAPIngestSegments, region_id));
+    uni_ps->write(std::move(del_batch), PageType::Local);
+}
+
+CheckpointIngestInfo::~CheckpointIngestInfo()
+{
+    if (clean_when_destruct)
+    {
+        removeFromPS();
+    }
+}
```

Comment:
should we add some `try ... catch ...` here? Or any exception thrown in dtor will make the process crash

Reply:
We'd better crash on any exception here. However, you are right, we should let it crash in a defined way